Release IPv6 address if IPv6 is disabled on an interface

robmry opened the pull request:
**- What I did**

When running:
```
  docker network create --ipv6 b46
  docker run --rm -ti \
    --network name=b46,driver-opt=com.docker.network.endpoint.sysctls=net.ipv6.conf.IFNAME.disable_ipv6=1 \
     busybox
```

IPv6 is enabled in the container and the network, so an IPv6 address will be allocated for the endpoint.

But, when the sysctl is applied, the IPv6 address will be removed from the interface ... so, no unsolicited neighbour advertisement should be (or can be) sent and, the endpoint should not be treated as dual-stack when selecting a gateway endpoint and, if it is selected as the gateway endpoint, setting up an IPv6 route via the network will fail.

In 27.x (the release that added per-endpoint sysctls), the error is ...

```
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running prestart hook #0: exit status 1, stdout: , stderr: failed to set IPv6 gateway while updating gateway: route for the gateway fd4b:b3cd:ce0c::1 could not be found: network is unreachable: unknown.
```

In 28.0 (because the `SetKey` hook isn't used any more, in this case), it's ...

```
docker: Error response from daemon: failed to set up container networking: failed to add interface veth3368cd8 to sandbox: failed to advertise addresses: write ip ::1->ff02::1: sendmsg: invalid argument
```

**- How I did it**

See the individual commit messages ... refactoring, with the bug fix in the final commit.

The refactoring means Sandbox and Endpoint config happens in the same order, whether or not the OS Sandbox has been created when the Endpoint is added.

**- How to verify it**

Exiting and new integration tests.

**- Description for the changelog**
```markdown changelog
- Fix an issue preventing container startup or selection of its network gateway when IPv6 is only disabled on a specific interface.
```

robmry committed:
Release IPv6 address if unused due to sysctl setting

When running:
  docker network create --ipv6 b46
  docker run --rm -ti \
    --network name=b46,driver-opt=com.docker.network.endpoint.sysctls=net.ipv6.conf.IFNAME.disable_ipv6=1 \
     busybox

IPv6 is enabled in the container and the network, so an IPv6 address
will be allocated for the endpoint.

But, when the sysctl is applied, the IPv6 address will be removed
from the interface ... so, no unsolicited neighbour advertisement
should be (or can be) sent and, the endpoint should not be treated
as dual-stack when selecting a gateway endpoint and, if it is
selected as the gateway endpoint, setting up an IPv6 route via the
network will fail.

So, if the IPv6 address disappears after sysctls have been applied,
release the address and remove it from the endpoint's config.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
> @robmry was this accidentally closed or superseded by the other PRs?

Oh - thank you! How do you manage to spot this stuff?!

I thought it was superseded by the unsolicited ARPs PR, but you prompted me to double-check - the last commit here is just stacked on that other change. I've forgotten where I got to with this one, but it probably needs resurrecting.

thaJeztah commented:
> How do you manage to spot this stuff?!


LOL, any activity gets in my notifications inbox; I tend to check up quickly on all of them, but also look for "this looks non-standard" things (like closing a PR); I didn't see a comment "why" it  was  closed, but I've had it myself that I accidentally pushed to the wrong PR branch, and GitHub automatically closing a PR, so thought "let me check!" ❤️

robmry committed:
Delay Endpoint config until the osSbox exists

When the SetKey hook is used (by a build container) it's called after
Endpoint.sbJoin, which will have called Sandbox.populateNetworkResources
to set up address, routes, sysctls and so on - but it's not able to do
any config until the osSbox exists. So, Sandbox.populateNetworkResources
is called again by SetKey to finish that config.

But, that means the rest of Endpoint.sbJoin has already happened before
the osSbox existed - it will have configured DNS, /etc/hosts, gateways
and so on before anything was set up for the OS.

So, if the osSbox configuration isn't applied as expected (for example,
a sysctl disables IPv6 on the endpoint), that sbJoin configuration is
incorrect.

To avoid unnecessary config+cleanup in thoses cases - delay the config
currently done by sbJoin until the osSbox exists.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Don't set up DNS in Network.createEndpoint

DNS is set up when the endpoint is joined to a network.

It was added in commit 4850c5f (Avoid duplicate entries in
/etc/hosts) then simplified in bcca214 (libnetwork: open-code
updating svc records) and seems to be related to setting up a
name on a swarm node that isn't running the container with the
endpoint.

But, all callers of Network.createEndpoint follow up with an
Endpoint.Join, which also sets up the DNS entry. Those callers are:
  Network.createLoadBalancerSandbox
  Network.CreateEndpoint
    - called by Daemon.connectToNetwork
    - called by Sandbox.setupDefaultGateway
    - called by builder-net/executor.go: iface.init

None of them bail out before the Join for a Swarm case.

So, it looks like enough has changed that the createEndpoint code is
no longer needed (it predates the internal DNS server) ... remove it.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Split OS-specific code out of Sandbox.populateNetworkResources

And move the Endpoint.populateNetworkResources code into the
all-platforms part of the Sandbox method.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Configure addresses before adding them to DNS or /etc/hosts

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
As this was already broken in 27.x, it fails with an error, and it's a corner case (disabling IPv6 on a single endpoint in an IPv6 network, rather than disabling it in the container) ... I've dropped the 28.0 milestone.

It's working, but I'll give it a check over before marking it as ready for review.

corhere reviewed (commented):
LGTM aside from the review comments.

corhere reviewed (approved):
LGTM!

akerouanton reviewed (approved):
One minor nit but feel free to ignore it. LGTM otherwise.

robmry commented on daemon/libnetwork/endpoint.go:
```diff
@@ -560,23 +560,43 @@ func (ep *Endpoint) sbJoin(ctx context.Context, sb *Sandbox, options ...Endpoint
 	}
 
-	// Current endpoint(s) providing external connectivity for the sandbox
+	// Current endpoint(s) providing external connectivity for the Sandbox.
+	// If ep is selected as a gateway endpoint once it's been added to the Sandbox,
+	// these are the endpoints that need to be un-gateway'd.
 	gwepBefore4, gwepBefore6 := sb.getGatewayEndpoint()
```
In the case where the old gateways are needed, they have to be looked up before `sb.addEndpoint(ep)` - because adding the endpoint might change the gateway selection. And, `addEndpoint` needs to be called when they're not needed.

So, it could be guarded with `sb.canPopulateNetworkResources()` ... but I'm not sure it's worth updating - in the common case, "canPopulate" is true, and in the buildkit case there's probably only one gateway endpoint to select from anyway (so `getGatewayEndpoint` is cheap).

I remember dithering about whether to just store the endpoint and early return - but decided to keep the store at the end, just in case the endpoint is updated by the populate/update calls, and those updates don't make it into the store.
